server/create.go

```diff
@@ -0,0 +1,628 @@
+			if err != nil {
+				ch <- gin.H{"error": err.Error()}
+			}
+		case map[string]any:
+			b, _ := json.Marshal(v) // re-marshal to JSON
+			if err := json.Unmarshal(b, &cfr); err == nil {
+				switch cfr.Type {
+				case "safetensors":
+					slog.Debug("create model from safetensors")
+					baseLayers, err = convertModelFromSafetensors(r, cfr, name, fn)
+					if err != nil {
+						ch <- gin.H{"error": err.Error()}
+						return
+					}
+				case "gguf":
+					slog.Debug("create model from gguf")
+					baseLayers, err = convertModelFromGGUF(r, cfr, name, fn)
+					if err != nil {
+						ch <- gin.H{"error": err.Error()}
+						return
+					}
+				default:
+					ch <- gin.H{"error": fmt.Sprintf("unknown type: %s", cfr.Type), "status": http.StatusBadRequest}
+					return
+				}
+			}
+		}
+
+		if err := createModel(r, name, baseLayers, fn); err != nil {
+			ch <- gin.H{"error": err.Error()}
+			return
+		}
+
+		if !envconfig.NoPrune() && oldManifest != nil {
+			if err := oldManifest.RemoveLayers(); err != nil {
+				ch <- gin.H{"error": err.Error()}
+			}
+		}
+
+		ch <- api.ProgressResponse{Status: "success"}
+	}()
+
+	if r.Stream != nil && !*r.Stream {
+		waitForStream(c, ch)
+		return
+	}
+
+	streamResponse(c, ch)
+}
+
+func convertModelFromSafetensors(r api.CreateRequest, cfr *api.CreateFromRequest, name model.Name, fn func(resp api.ProgressResponse)) ([]*layerGGML, error) {
+	tmpDir, err := os.MkdirTemp("", "ollama-safetensors")
+	if err != nil {
+		return nil, err
+	}
+	defer os.RemoveAll(tmpDir)
+
+	for _, file := range cfr.Files {
+		blobPath, err := GetBlobsPath(file.Digest)
+		if err != nil {
+			return nil, err
+		}
+		if err := createLink(blobPath, filepath.Join(tmpDir, file.Path)); err != nil {
+			return nil, err
+		}
+	}
+
+	t, err := os.CreateTemp(tmpDir, "fp16")
+	if err != nil {
+		return nil, err
+	}
+	defer t.Close()
+	fn(api.ProgressResponse{Status: "converting the model"})
+
+	if err := convert.ConvertModel(os.DirFS(tmpDir), t); err != nil {
+		return nil, err
+	}
+
+	if _, err := t.Seek(0, io.SeekStart); err != nil {
+		return nil, err
+	}
+
+	layer, err := NewLayer(t, "application/vnd.ollama.image.model")
+	if err != nil {
+		return nil, err
+	}
+
+	bin, err := layer.Open()
+	if err != nil {
+		return nil, err
+	}
+
+	ggml, _, err := llm.DecodeGGML(bin, 0)
+	if err != nil {
+		return nil, err
+	}
+	layers := []*layerGGML{{layer, ggml}}
+
+	return detectChatTemplate(layers)
+}
+
+func convertModelFromGGUF(r api.CreateRequest, cfr *api.CreateFromRequest, name model.Name, fn func(resp api.ProgressResponse)) ([]*layerGGML, error) {
+	if len(cfr.Files) == 0 {
+		return nil, fmt.Errorf("no files provided")
+	} else if len(cfr.Files) > 1 {
+		return nil, fmt.Errorf("only one file is supported")
+	}
+
+	file := cfr.Files[0]
+	return getGGUFLayers(file, fn)
+}
+
+func createModel(r api.CreateRequest, name model.Name, baseLayers []*layerGGML, fn func(resp api.ProgressResponse)) (err error) {
+	// todo move to the new config type
```

Comment:
What is the new config type?

Reply:
I'm going to remove the comment. The idea is we could create our own manifest format which would probably include the config instead of storing it as a separate layer.